lib/AccessToMemory/Path.php

```
@@ -0,0 +1,175 @@
+            $items[] = new Path($this->path.DIRECTORY_SEPARATOR.$item);
+        }
+
+        return $items;
+    }
+
+    /**
+     * Remove a directory.
+     *
+     * @param bool $recursive delete a directory and all its contents when true
+     */
+    public function rmdir($recursive = false)
+    {
+        $this->throwExceptionIfNotADir();
+
+        if ($recursive) {
+            foreach ($this->ls() as $child) {
+                $child->delete($recursive);
+            }
+        }
+
+        if (!empty($this->ls())) {
+            throw new \RuntimeException(
+                sprintf(
+                    "Can't delete %s: directory is not empty",
+                    $this->path
+                )
+            );
+        }
+
+        rmdir($this->path);
+    }
+
+    /**
+     * Delete a file.
+     */
+    public function unlink()
+    {
+        $this->throwExceptionIfNotAFile();
+
+        unlink($this->path);
+    }
+
+    /**
+     * Delete a file or a directory.
+     *
+     * @param bool $recursive if true, delete a directory and all its contents
+     */
+    public function delete($recursive = false)
+    {
+        if (is_dir($this->path)) {
+            $this->rmdir($recursive);
+        } else {
+            $this->unlink();
+        }
+    }
+
+    protected function throwExceptionIfFileNotFound()
+    {
+        if (!$this->exists()) {
+            throw new \RuntimeException(
+                sprintf('File not found: %s', $this->path)
+            );
+        }
+    }
+
+    protected function throwExceptionIfNotADir()
+    {
+        if (!$this->isDir()) {
+            throw new \RuntimeException(
+                sprintf('%s is not a directory', $this->path)
+            );
+        }
+    }
+
+    protected function throwExceptionIfNotAFile()
+    {
+        if (!$this->isFile()) {
+            throw new \RuntimeException(
+                sprintf('%s is not a file', $this->path)
+            );
+        }
+    }
+}
```

You went deep in this abstraction to just delete the directory, hopefully we give it more use in the future.

---

Yes, I also needed to recursively delete a directory in https://github.com/artefactual/atom/pull/1500.  

In the PR1500 I used [sfToolkit::clearDirectory()](https://github.com/artefactual/atom/blob/938ee91eeff42c168aa82c641b7e979785596634/lib/job/arZipFileDownload.class.php#L84) which I copied from the previous code. When I was working on this fix I examined `sfToolkit::clearDirectory()` and realized that it won't delete a directory that contains `.svn` or `CVS` files anywhere in it's subtree. :(   It's pretty unlikely that a DIP or export directory would have a CVS or .svn file, but it seems to me that it's a non-intuitive behaviour that could be a problem in some cases.

I also found in writing the Path library that the PHP functions like "rmdir()" an "unlink()" will fail without any error if e.g. you try and `unlink()` a directory, or vice versa.   I think some times a silent failure is fine, and other times it creates a bug. I think it's better in general to consistently throw an exception on failure, and give the developer the choice to catch the exception if they don't want the script to halt.